gRPC changes for FSMaster services

ggezer opened the pull request.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/984/
Test FAILed.

madanadit commented on core/transport-options/src/grpc/file_system_master_options.proto:
```diff
@@ -0,0 +1,172 @@
+syntax = "proto2";
```
we should move to proto3 soon

ggezer replied:
It's on my task list for the consolidation work after security.

ggezer replied:
Looks like proto3 dropped support for checking presence of a value, so no more protoOption.hasQwe() family of methods. 
@madanadit , @aaudiber WDYT?

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/1015/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/1043/
Test FAILed.

ggezer commented:
@madanadit, @aaudiber PTAL.

**High level summary of changes:**
- Remove option wrappers on client side.
- Replace option wrappers on master side with OperationContext implementations.
- Implement merging of master options with given options. So that going forward clients don't need to send defaults. (Client side is not modified yet to not send defaults)

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/1046/
Test FAILed.

aaudiber commented on core/client/fs/src/main/java/alluxio/client/ReadType.java:
```diff
@@ -82,4 +84,15 @@ public boolean isCache() {
   }
+
+  // TODO(ggezer) Eliminate below parsers/converters for ReadType
+  private static final String sProtoNameHeader = "READ_";
```
why is this needed?

ggezer replied:
gRPC enums are defined in a global scope. Defining them without some kind of header invites conflicts for commonly used enum values. (E.g NONE,ALL, etc...)

ggezer replied:
PS: I don't intend to keep this header trick for longer. Once we start eliminating wire types, they should go away naturally.

aaudiber commented on core/client/fs/src/main/java/alluxio/client/file/FileSystemClientOptions.java:
```diff
@@ -0,0 +1,234 @@
+ * The file system class to set default options for client.
+ */
+@ThreadSafe
+public final class FileSystemClientOptions {
```
Is this temporary? I thought we were going to fill in defaults on the server side

ggezer replied:
Yes, but depends.
It's temporarily used here to provide client defaults. (Only the infra is done but the clients are still supplying defaults)

However, if there are discrepancies between client/master default values, then we would need to use this to merge client provided settings with client defaults, which would ultimately be merged with master defaults.

I'll tackle getting rid of client defaults as a separate work.

aaudiber commented on core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java:
```diff
@@ -0,0 +1,108 @@
+  private long mOperationTimeMs;
+  private UfsStatus mUfsStatus;
+
+  // Prevent instantiation
```
this is already accomplished by the other private constructor

aaudiber commented on core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java:
```diff
@@ -0,0 +1,108 @@
+   * @return the instance of {@link CompleteFileContext} with default values for master
+   */
+  public static CompleteFileContext defaults(CompleteFilePOptions.Builder optionsBuilder) {
+    CompleteFilePOptions masterOptions = FileSystemMasterOptions.getCompleteFileOptions();
```
naming is confusing here - `FileSystemMasterOptions.getCompleteFileOptions()` gets default option values but doesn't mention anything about defaults, while `CompleteFileContext.defaults(builder)` is named `defaults`, but returns non-default values

One idea:
`defaults` -> `applyDefaults`
`FileSystemMasterOptions.getCompleteFileOptions` -> `FileSystemMasterOptions.completeFileDefaults()`

aaudiber replied:
we could also move the `FileSystemMasterOptions.getCompleteFileOptions` to be `private static CompleteFileContext.defaults()`. I don't think it needs to be exposed outside of `CompleteFileContext`

ggezer replied:
Ultimate plan is to get rid of FileSystemMasterOptions and move the definitions to corresponding operation contexts. I'll do that all at once for all operation types.

ggezer replied:
After reconsideration I feel like *Context.defaults() overload for taking proto options makes sense. Parameter javadoc also clearly states that given options are due for merging with defaults.

aaudiber replied:
I still think it's misleading. We have historically used `.defaults()` methods on all of our option classes to return the default options. How about `CompleteFileContext.create(completeFileOptions)`?

aaudiber commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -803,17 +804,18 @@ public FileInfo getFileInfo(AlluxioURI path, GetStatusOptions options)
+            LoadMetadataContext.defaults(LoadMetadataPOptions.newBuilder().setCreateAncestors(true)
+                .setCommonOptions(FileSystemMasterCommonPOptions.newBuilder()
+                    .setTtl(context.getOptions().getCommonOptions().getTtl())
+                    .setTtlAction(context.getOptions().getCommonOptions().getTtlAction()))));
```
probably worth saving `context.getOptions()` and `context.getOptions.getCommonOptions()` to local variables since they are used many times.

madanadit commented on bin/alluxio:
```diff
@@ -398,9 +398,13 @@ function main {
     rm -rf ${BIN}/../core/transport/src/main/java/alluxio/grpc
     for src_file in $(find ${BIN}/../core/transport/src/grpc -name '*.proto' -type f); do
-      protoc --java_out=${BIN}/../core/transport/src/main/java --proto_path=${BIN}/../core/transport/src/grpc --plugin=protoc-gen-grpc-java=${BIN}/protoc-gen-grpc-java-1.10.1-osx-x86_64.exe --grpc-java_out=${BIN}/../core/transport/src/main/java ${src_file}
+      protoc --java_out=${BIN}/../core/transport/src/main/java --proto_path=${BIN}/../core/transport/src/grpc --proto_path=${BIN}/../core/transport-options/src/grpc --plugin=protoc-gen-grpc-java=${BIN}/protoc-gen-grpc-java-1.10.1-osx-x86_64.exe --grpc-java_out=${BIN}/../core/transport/src/main/java ${src_file}
```
is there a better way than referencing `${BIN}/protoc-gen-grpc-java-1.10.1-osx-x86_64.exe`?

ggezer replied:
Making maven to generate the files maybe?

madanadit replied:
i would prefer pre-generated files. we can defer this to following PRs

ggezer replied:
I didn't mean generating during the build but invoking the generation phase only when we want to regenerate, not triggering the generation in the regular build.

ggezer replied:
I put this on my task list for the consolidation work.

madanadit commented on core/client/fs/src/main/java/alluxio/client/AlluxioStorageType.java:
```diff
@@ -11,6 +11,9 @@
 
 package alluxio.client;
 
+import alluxio.grpc.WritePType;
```
both these are not used?

madanadit commented on core/client/fs/src/main/java/alluxio/client/UnderStorageType.java:
```diff
@@ -11,6 +11,9 @@
 
 package alluxio.client;
 
+import alluxio.grpc.WritePType;
```
same here

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/1050/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/1055/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/1059/
Test FAILed.

aaudiber commented on core/server/master/src/main/java/alluxio/master/file/options/CompleteFileContext.java:
```diff
@@ -0,0 +1,104 @@
+   * @param optionsBuilder Builder for proto {@link CompleteFilePOptions} to merge with defaults
+   * @return the instance of {@link CompleteFileContext} with default values for master
+   */
+  public static CompleteFileContext defaults(CompleteFilePOptions.Builder optionsBuilder) {
```
Can we call this `create`, `applyDefaults`, or `createWithDefaults`? We typically use verbs for methods that process an input.

ggezer replied:
Agreed to naming.

Considering the overall comment you gave, on keeping the given defaults as is, how about keeping the proto options as builder but provide another method from base OperationContext to expose given defaults. So in the end we would have above three methods for creating the contexts and each context would have getOriginalOptions() method that returns whatever was provided by the caller before merging or modification. 

Using a builder makes setting proto fields very easy. It's not just to avoid extra .build() call at the end but to be able to set proto fields line-by-line as below, which is very usefull in places where there are many proto fields (e.g. InodeTree).

context = SomeContext.createWithDefaults();
context.getOptions().setQwe();
context.getOptions().setAsd();

compared to;

context = SomeContext.createWithDefaults();
context.setOptions(context.getOptions().toBuilder().setQwe().setAsd().build);

aaudiber replied:
At the start of the RPC we receive options from the client and merge them with the defaults to create a complete `*POptions` object. Most RPCs never modify this `*POptions` object. Ideally no RPCs would modify it. We can guarantee this by making it immutable, simplifying our lives greatly because we don't need to worry about all the options changing.

Modifying options during an RPC call is rare. I only see three places where we need to do it:

https://github.com/Alluxio/alluxio/pull/8122/files#diff-7f82122f923de058e3c4a20b43cd7e4eR807
https://github.com/Alluxio/alluxio/pull/8122/files#diff-7f82122f923de058e3c4a20b43cd7e4eR890
https://github.com/Alluxio/alluxio/pull/8122/files#diff-7f82122f923de058e3c4a20b43cd7e4eR921

All 3 of these are related to loadMetadataType/loadDescendantType, which are effectively deprecated now that we have ufs sync. We shouldn't base the convention off of these outliers. Immutability is valuable for preventing bugs, and we have an opportunity here to make dozens of fields (every option) immutable at a low cost code-wise

ggezer replied:
@madanadit, this discussion is related to your question :
> should we set all options fields in one go?

ggezer replied:
I put this on my task list for the consolidation work, as it'll cause lots of merge conflicts for pending PRs.

madanadit commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -2451,37 +2454,42 @@ private void loadFileMetadataInternal(RpcContext rpcContext, LockedInodePath ino
-        .setGroup(ufsStatus.getGroup());
-    short ufsMode = ufsStatus.getMode();
+    CreateFileContext createFileContext = CreateFileContext.defaults();
+    createFileContext.getOptions().setBlockSizeBytes(ufsBlockSizeByte);
```
why set createFileContext instead of the way it was?

madanadit replied:
i see. should we set all `options` fields in one go?

ggezer replied:
We didn't have to. However, due to discussions for immutable of options, that's what's going to happen. So yes, we should set it all in one go.

madanadit commented on core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java:
```diff
@@ -2451,37 +2454,42 @@ private void loadFileMetadataInternal(RpcContext rpcContext, LockedInodePath ino
+              .setUfsStatus(context.getUfsStatus());
       if (ufsLastModified != null) {
-        completeOptions.setOperationTimeMs(ufsLastModified);
+        completeCtx.setOperationTimeMs(ufsLastModified);
```
nit: consistent w/ naming => `completeContext`

madanadit commented on core/server/master/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java:
```diff
@@ -14,17 +14,25 @@
 import alluxio.AlluxioURI;
 import alluxio.Constants;
 import alluxio.RestUtils;
+import alluxio.client.file.FileSystemClientOptions;
```
why client options default?

madanadit commented on core/server/master/src/main/java/alluxio/master/file/InodeTtlChecker.java:
```diff
@@ -98,9 +101,10 @@ public void heartbeat() {
+                      DeleteContext.defaults(DeletePOptions.newBuilder().setRecursive(true)));
                 } else {
-                  mFileSystemMaster.delete(path, DeleteOptions.defaults().setRecursive(false));
+                  mFileSystemMaster.delete(path, DeleteContext.defaults());
```
is this intentional?

madanadit commented on core/server/master/src/main/java/alluxio/master/file/meta/InodeDirectory.java:
```diff
@@ -17,12 +17,13 @@
 import alluxio.exception.InvalidPathException;
 import alluxio.master.ProtobufUtils;
-import alluxio.file.options.CreateDirectoryOptions;
+import alluxio.master.file.options.CreateDirectoryContext;
```
should we rename the package to `context`?

ggezer commented on core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java:
```diff
@@ -145,26 +150,27 @@ public FileOutStream createFile(AlluxioURI path, CreateFileOptions options)
-    outStreamOptions.setAcl(status.getAcl());
+
     try {
+      OutStreamOptions outStreamOptions = new OutStreamOptions(options);
```
Because now OutStream is created by a constructor that takes proto options during which we try to create a class instance using class name. 

I wrapped that code in OutStreamOptions with try-catch and throw RuntimeException. Code is out of the try-catch now.

ggezer commented on core/client/fs/src/main/java/alluxio/client/file/options/InStreamOptions.java:
```diff
@@ -11,52 +11,81 @@
     mStatus = status;
-    mOptions = options;
+    mProtoOptions = options;
+    CreateOptions blockLocationPolicyCreateOptions = CreateOptions.defaults()
```
JobUtils.loadBlock also calls that along with many other tests.

ggezer commented on core/client/fs/src/test/java/alluxio/client/block/AlluxioBlockStoreTest.java:
```diff
@@ -94,11 +95,19 @@
    */
   @ThreadSafe
-  private static class MockFileWriteLocationPolicy
+  public static class MockFileWriteLocationPolicy
```
I thought it was necessary now that creating an instance of it was done by Class.forName within the InStreamOptions. I tried private and it worked. Reverted it.

ggezer commented on core/server/common/src/main/java/alluxio/master/AbstractNonJournaledMaster.java:
```diff
@@ -46,6 +47,6 @@ public void resetState() {
   @Override
   public Iterator<JournalEntry> getJournalEntryIterator() {
-    return Iterators.emptyIterator();
+    return Iterators.unmodifiableIterator(Collections.emptyIterator());
```
Due to Guava version bump, emptyIterator() is no longer there. This is a functional equivalent of it.

ggezer commented on core/client/fs/src/main/java/alluxio/client/file/BaseFileSystem.java:
```diff
@@ -145,26 +150,27 @@ public FileOutStream createFile(AlluxioURI path, CreateFileOptions options)
     } catch (Exception e) {
       delete(path);
-      throw e;
+      throw new AlluxioException(e.getMessage(), e);
```
You can check the public constructor for OutStreamOptions, and there is now a Class.forName() call. I wrapped that code with try-catch and reverted this file to its original.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/1128/
Test FAILed.

ggezer commented:
@madanadit @aaudiber. Updated the PR. PTAL.

FYI, I postponed addressing some feedbacks to the a later PR in below schedule.
 1- FSMaster gRPC migration
 2- Rest of service migrations (BlockMaster/JobMaster/MetaMaster/KeyValueMaster)
 3- Security
 4- Consolidation (removing wire types/ operation context changes/ general improvement and cleaning up of code)
